Internal release: record stat to track how often unverified domain email nag is shown

mpkelly opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes n/a
Related https://github.com/Automattic/jetpack/pull/33390

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Record a stat for how many times email nag added in https://github.com/Automattic/jetpack/pull/33390 is displayed

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Apply the patch to your sandbox as per this comment https://github.com/Automattic/jetpack/pull/33855#issuecomment-1785044907
* Sandbox the wordpress.com subdomain for your site. 
* Follow the test plan here to get a site with an unverified email https://github.com/Automattic/wp-calypso/pull/83597
* Go to the site's front end and confirm a nah is shown. Check bump stat `wpcom_frontend_unverified_domain_email_nag_shown` and see that the stat has increased (there might be a delay)
* Is the stat name I used appropriate?

mpkelly committed:
Add bump stat for unverified domain email nag

mpkelly committed:
changelog

mpkelly committed:
Update suspended-domains-add-bump-stats-for-nag

mpkelly committed:
Use direct stats URL to avoid depending on function being available at runtime

danielbachhuber commented on projects/packages/jetpack-mu-wpcom/src/features/domain-email-nag/domain-email-nag.php:
```diff
@@ -49,6 +49,20 @@ function get_domain_with_unverified_email() {
+		'wpcom-domain-email-nag-shown',
+		$domain
+	);
+	wp_remote_get( $pixel_url );
```
`wp_remote_get()` is blocking and will slow down the request. Could we display the pixel instead?

mpkelly replied:
I will switch to rendering an image instead. Thanks!

p-jackson commented on projects/packages/jetpack-mu-wpcom/src/features/domain-email-nag/domain-email-nag.php:
```diff
@@ -49,6 +49,20 @@ function get_domain_with_unverified_email() {
+function bump_mc_stats( $domain ) {
+	$pixel_url = sprintf(
+		'http://pixel.wp.com/g.gif?v=wpcom-no-pv&x_%s=%s',
+		'wpcom-domain-email-nag-shown',
```
I think we need to use pascal_case instead of kebab-case. Pretty sure invalid names get filtered out.
PCYsg-4S2-p2

p-jackson replied:
Maybe this?
`wpcom_frontend_unverified_domain_email_nag_shown`

I like saying frontend because it wasn't initially clear to me what context the nag was shown in.

p-jackson commented on projects/packages/jetpack-mu-wpcom/src/features/domain-email-nag/domain-email-nag.php:
```diff
@@ -77,7 +93,7 @@ function domain_email_nag() {
 			<a class="button" href="<?php echo esc_url( get_account_url() ); ?>"><?php esc_html_e( 'Fix', 'jetpack-mu-wpcom' ); ?></a>
 		</div>
-	</div>
+	</div>	
```
whitespace nitpick

p-jackson reviewed (commented):
Made some notes but I wasn't able to test this unfortunately.
I have an unverified email, and I can see the unverified domain email banner in the Calypso sidebar, but after sandboxing the frontend and applying this diff to my sandbox, I saw no banner 😕

mpkelly commented:
> * Sandbox the wordpress.com subdomain for your site.

@p-jackson, did you do this part? It would mean an entry in your `/etc/hosts` like `192.0.94.44 testname.wordpress.com`

mpkelly committed:
Use image URL to record stat to avoid blocking API call; tidy whitespace; update name of stat.

p-jackson commented on projects/packages/jetpack-mu-wpcom/src/features/domain-email-nag/domain-email-nag.php:
```diff
@@ -63,6 +63,12 @@ function domain_email_nag() {
 	}
 
+	$pixel_url = sprintf(
+		'http://pixel.wp.com/g.gif?v=wpcom-no-pv&x_%s=%s',
```
I prefer using `b.gif` which does the same thing but renders a transparent pixel rather than a smiley face. I'm always paranoid that there'll be some error that causes the image to appear really big or something, and the smiley face would feel inappropriate then 😅 that's me being paranoid though.

I'm not seeing any of these stats appear in MC yet. I would have thought I'd see something from testing (even if I personally have only just managed to get the nag showing).

Maybe that's because we need to change the URL to something like. `&x_wpcom_frontend_unverified_domain_email_nag=shown`, with an equals in there. I don't see why the group and event name can't be the same. But maybe it can't and I just remembered that this way of doing things (separating out the verb into a separate name) is more common. So maybe it's required or something.

mpkelly replied:
TIL. Thanks for this change.

p-jackson reviewed (commented):
I was able to figure it out why I wasn't seeing the nag. It's because `class_exists( 'Domain_Management' )` returns `false` for me.

I can see why. `Domain_Management` is defined in `admin-plugins`, and admin plugins are only loaded when in the admin view. We're rendering the front end here and so I'd totally expect this class not to exist.

I tested the stat by requiring the `Domain_Management` code on my sandbox and I got the nag appearing. But I don't think it's good to do that in production. We keep front end rendering performant by not loading all the code, and who knows what spaghetti would get included if we started to require it.

I'll merge this once I see the event working in MC. But I think we'll need a follow up which finds a way to get icann info on the front end. Maybe it'll be ok as long as we only load code when we know it's the owner who's logged in 🤔

p-jackson committed:
Replace g.gif with b.gif

Change stats pixel from smiley picture to transparent pixel.

p-jackson committed:
Anonymise the domain email nag stat count

Switched to recording the "shown" verb as the actual event, not the
site's domain name. An advantage of b.gif stats is that they are
truly anonymous.

p-jackson reviewed (dismissed):
I discovered a way to test this change.

I've made some changes to use `b.gif` instead of `g.gif` and also anonymised the bump stat. I actually wonder whether using the domain as the event name might have caused issues with the URL not being encoded correctly.

mpkelly commented:
> I discovered a way to test this change.
> 

I am just testing it now and no longer see the banner. I can't see why, but I will take a closer look tomorrow. 

Edit. Been a long day. There is a discussion on Slack about removing the dependency on `Domain_Management` which I think should make this banner show consistently.

mpkelly committed:
Rework nag logic to be async and based on wc_enqueue_js

mpkelly committed:
Merge branch 'suspended-domains/add-bump-stats-for-nag' of https://github.com/Automattic/jetpack into suspended-domains/add-bump-stats-for-nag

mpkelly committed:
negate domain check

mpkelly commented:
I've updated this but haven't tested it yet. I will do that tomorrow (8th)

p-jackson commented on projects/packages/jetpack-mu-wpcom/src/features/domain-email-nag/domain-email-nag.php:
```diff
@@ -32,47 +32,69 @@ function get_account_url() {
-	if ( ! $domain ) {
-		return;
-	}
+	$blog_id = Jetpack_Options::get_option( 'id' );
```
This one's causing an error on the front end
![CleanShot 2023-11-09 at 20 10 55@2x](https://github.com/Automattic/jetpack/assets/1500769/b6c7cd7b-bfe1-4538-9b55-8432fbff3259)

I bet it's a missing namespace.

p-jackson commented on projects/packages/jetpack-mu-wpcom/src/features/domain-email-nag/domain-email-nag.php:
```diff
@@ -32,47 +32,69 @@ function get_account_url() {
+/**
+ * Checks if the site is using a custom domain.
+ */
+function is_custom_domain() {
```
This `is_custom_domain` check means we'll only perform the check if the primary domain is custom. I found myself in a situation where my primary address was `*.wordpress.com` but I have a secondary domain which is custom. This check means I don't see the nag even though the API is supposed to be checking all my domains, not just the primary one.

I like the optimisation of trying to only enqueue the nag code when it's necessary, but I don't think we enough information here 🧐

mpkelly replied:
I still prefer this as doing an API call every time the user is on the home feels too expensive to me, and I think most people will use their custom domain instead of ours.

p-jackson replied:
I guess if they're still using `.wordpress.com` as their primary address it's not as pressing an issue to verify their domain email. It won't break their site to the same extent.

p-jackson commented on projects/packages/jetpack-mu-wpcom/src/features/domain-email-nag/domain-email-nag.php:
```diff
@@ -32,47 +32,69 @@ function get_account_url() {
+ */
+function is_custom_domain() {
+	$domain = get_domain();
+	return ! str_ends_with( $domain, 'wpcomstaging.com' ) && ! str_ends_with( $domain, 'wordpress.com' );
```
```suggestion
	return ! str_ends_with( $domain, '.wpcomstaging.com' ) && ! str_ends_with( $domain, '.wordpress.com' );
```

We should have the extra dots in the substring check. Otherwise it'll return `false` if the user does something clever like register the `my-special-domain-wordpress.com` domain.

The security workshop has made me paranoid 😅

mpkelly replied:
Yeah, I would still prefer this as doing an API call every time the user is on the home feels too expensive to me, and I think most people will use their custom domain instead of ours.

p-jackson commented on projects/packages/jetpack-mu-wpcom/src/features/domain-email-nag/domain-email-nag.php:
```diff
@@ -32,47 +32,69 @@ function get_account_url() {
+	$blog_id = Jetpack_Options::get_option( 'id' );
 
-	wp_enqueue_style( 'wpcom-domain-email-nag-style', plugins_url( 'domain-nag.style.css', __FILE__ ), array(), Jetpack_Mu_Wpcom::PACKAGE_VERSION );
+	wc_enqueue_js(
```
`wc_enqueue_js` is defined in Woo so pretty sure we can't use it here.

`wc_enqueue_js` does its magic by subscribing to the `wp_footer` action and then echoing out the `<script>` tag. We can probably just do that ourselves.

mpkelly replied:
Thanks for that. I will just use a script tag.

p-jackson commented on projects/packages/jetpack-mu-wpcom/src/features/domain-email-nag/domain-email-nag.php:
```diff
@@ -32,47 +32,69 @@ function get_account_url() {
+				}
+			});
+		",
+			$blog_id
```
Seeing the variable injected directly into a JavaScript string made me feel nervous. Even though I'm not sure how a user could spoof it. Could we do something like this to be extra paranoid?

```
filter_var( $blog_id, FILTER_VALIDATE_INT )
```

There might also be a core WP function for this sort of thing.

p-jackson commented on projects/packages/jetpack-mu-wpcom/src/features/domain-email-nag/domain-email-nag.php:
```diff
@@ -32,47 +32,69 @@ function get_account_url() {
+		sprintf(
+			"
+			const base = 'https://public-api.wordpress.com';
+			const path = '/v1/domains/%s/is-domain-email-unverified';
```
Similar to the above suggestion, maybe it's enough to do this?

```suggestion
			const path = '/v1/domains/%d/is-domain-email-unverified';
```

But I worry that it'll throw an error at runtime instead of fail silently.

p-jackson reviewed (commented):
Made some notes below. Also making a note that we should make sure to check this on atomic too before deploying. It's in the mu plugin so it runs on atomic too.

Another idea, probably for another diff, but I saw this UI in the launcher header (I'm not sure what we call it). Maybe this'd be a more attractive place for the nag to appear?

![CleanShot 2023-11-09 at 19 58 42@2x](https://github.com/Automattic/jetpack/assets/1500769/22b62e64-9d82-49bb-a4b0-09dbb525bc23)

mpkelly committed:
Update projects/packages/jetpack-mu-wpcom/src/features/domain-email-nag/domain-email-nag.php

Co-authored-by: Philip Jackson <p-jackson@users.noreply.github.com>

mpkelly committed:
Update projects/packages/jetpack-mu-wpcom/src/features/domain-email-nag/domain-email-nag.php

Co-authored-by: Philip Jackson <p-jackson@users.noreply.github.com>

mpkelly committed:
Fix some issues with dependencies; validate blog_id

mpkelly committed:
Some updates to get this working:
- use heredoc
- dont escape output
- use https for stas

mpkelly commented:
It's been a bit of a fight but this is getting closer. I had to make some local modifications when testing because I am using a test store domain and these don't get included in the API response. I will need to buy a real domain to test this before merging. Still need to test on Atomic too. I also ran into a permission issue so need to debug the backend diff. 

![image](https://github.com/Automattic/jetpack/assets/6851384/dd16f945-2980-4a18-a6eb-38ae7e3a607f)

mpkelly committed:
Update path after reworking backend

mpkelly committed:
Add domains!

p-jackson commented on projects/packages/jetpack-mu-wpcom/src/features/domain-email-nag/domain-email-nag.php:
```diff
@@ -32,52 +32,80 @@ function get_account_url() {
+
+	$script = sprintf(
+		$script,
+		$blog_id
```
Just a heads up that this `$blog_id` value hasn't been run through the `filter_var` function.

mpkelly replied:
The way I did it was to return early if it is not a valid `int`. Should be ok like this?

p-jackson replied:
Oh I see your right. For some reason I remember `filter_var()` always returning a valid int even if you passed it garbage. But it actually returns `false`, so this is fine.

p-jackson commented on projects/packages/jetpack-mu-wpcom/src/features/domain-email-nag/domain-email-nag.php:
```diff
@@ -32,52 +32,80 @@ function get_account_url() {
 			<a class="button" href="<?php echo esc_url( get_account_url() ); ?>"><?php esc_html_e( 'Fix', 'jetpack-mu-wpcom' ); ?></a>
 		</div>
 	</div>
+	<script><?php echo $script; // phpcs:ignore -- output generated on server ?></script>
```
I like this. Sticking a `<script>` right here is just simpler. I think we can add the `defer` attribute so the JS isn't render blocking and it behaves more like the script appearing in the footer.

mpkelly commented:
I'm getting a 403 now, so will need to send over a token with the request. Otherwise, I think this is ready.

mpkelly commented:
This proved quite challenging and has performance implications for all sites with custom domains, so we figured it was not worth it. Closing.
